src/envoy/http/authn/authn_utils.cc

```diff
@@ -0,0 +1,62 @@
+/* Copyright 2018 Istio Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "authn_utils.h"
+#include "src/envoy/http/jwt_auth/jwt_authenticator.h"
+
+namespace Envoy {
+namespace Http {
+namespace Istio {
+namespace AuthN {
+
+bool AuthnUtils::GetJWTPayloadFromHeaders(
+    const HeaderMap& headers,
+    ::google::protobuf::Map< ::std::string, ::std::string>* claims) {
+  const HeaderEntry* entry =
+      headers.get(JwtAuth::JwtAuthenticator::JwtPayloadKey());
+  if (!entry) {
+    return false;
+  }
+  std::string value(entry->value().c_str(), entry->value().size());
+  std::string payload_str = JwtAuth::Base64UrlDecode(value);
+  // Return an empty string if Base64 decode fails.
+  if (payload_str.empty()) {
+    ENVOY_LOG(error, "Invalid {} header, invalid base64: {}",
+              JwtAuth::JwtAuthenticator::JwtPayloadKey().get(), value);
+    return false;
+  }
+  try {
+    auto json_obj = Json::Factory::loadFromString(payload_str);
+    json_obj->iterate(
+        [claims](const std::string& key, const Json::Object& obj) -> bool {
```

Comment:
I feel this is missing smt. I know this is copied from CheckData, but the jwt implementation to extract payload is like this (note the special treatment for "aud"), and afaik, that payload is the same payload the jwt filter put back to the output (it simply strip the header and signature off the original jwt). Please check with @qiwzhang  on this.

https://github.com/istio/proxy/blob/0e2a8fc267887c3f5862fd202edc35720cef55c3/src/envoy/http/jwt_auth/jwt.cc#L312

Reply:
This missing state for handling the exception is intentional. When (*claims)[key] = obj.asString() is called, if the obj type is not string, an exception is thrown. But this should not be treated as error (i.e., return false) since other objects still need to be added to the claims.

Reply:
'aud' could be array.

Reply:
Added the handling of the case when 'aud' may be an array of strings or a single string.